jobs/mysql_backup_job.go

```diff
@@ -0,0 +1,190 @@
+		name:     name,
+		dsn:      dsn,
+		location: locationConfig,
+	}
+}
+
+// ID returns job id.
+func (j *MySQLBackupJob) ID() string {
+	return j.id
+}
+
+// Type returns job type.
+func (j *MySQLBackupJob) Type() string {
+	return "mysql_backup"
+}
+
+// Timeout returns job timeout.
+func (j *MySQLBackupJob) Timeout() time.Duration {
+	return j.timeout
+}
+
+func (j *MySQLBackupJob) Run(ctx context.Context, send Send) error {
+	t := time.Now()
+	j.l.Info("MySQL backup started")
+
+	mysqlConfig, err := mysql.ParseDSN(j.dsn)
+	if err != nil {
+		return errors.Wrapf(err, "mysql parse dsn")
+	}
+
+	if _, err := exec.LookPath(xtrabackupBin); err != nil {
+		return errors.Wrapf(err, "lookpath: %s", xtrabackupBin)
+	}
+
+	if j.location.S3Config != nil {
+		if _, err := exec.LookPath(xbcloudBin); err != nil {
+			return errors.Wrapf(err, "lookpath: %s", xbcloudBin)
+		}
+	}
+
+	xtrabackupCmd := exec.CommandContext(ctx, xtrabackupBin,
+		"--user="+mysqlConfig.User,
+		"--password="+mysqlConfig.Passwd,
+		"--compress",
+		"--backup") // #nosec G204
+
+	switch mysqlConfig.Net {
+	case "tcp":
+		splitAddr := strings.Split(mysqlConfig.Addr, ":")
+		xtrabackupCmd.Args = append(xtrabackupCmd.Args, "--host="+splitAddr[0])
+		if len(splitAddr) > 1 {
+			xtrabackupCmd.Args = append(xtrabackupCmd.Args, "--port="+splitAddr[1])
+		}
+	case "unix":
+		xtrabackupCmd.Args = append(xtrabackupCmd.Args, "--socket="+mysqlConfig.Addr)
+	}
+
+	var xbcloudCmd *exec.Cmd
+	switch {
+	case j.location.S3Config != nil:
+		xtrabackupCmd.Args = append(xtrabackupCmd.Args, "--stream=xbstream")
+		xbcloudCmd = exec.CommandContext(ctx, xbcloudBin,
+			"put",
+			"--storage=s3",
+			"--s3-endpoint="+j.location.S3Config.Endpoint,
+			"--s3-access-key="+j.location.S3Config.AccessKey,
+			"--s3-secret-key="+j.location.S3Config.SecretKey,
+			"--s3-bucket="+j.location.S3Config.BucketName,
+			"--s3-region="+j.location.S3Config.BucketRegion,
+			"--parallel=10",
+			j.name) // #nosec G204
+	default:
+		return errors.Errorf("unknown location config")
+	}
+
+	var outBuffer bytes.Buffer
+	var errBackupBuffer bytes.Buffer
+	var errCloudBuffer bytes.Buffer
+	xtrabackupCmd.Stderr = &errBackupBuffer
+
+	xtrabackupStdout, err := xtrabackupCmd.StdoutPipe()
+	if err != nil {
+		return errors.Wrapf(err, "failed to get xtrabackup stdout pipe")
+	}
+
+	wrapError := func(err error) error {
+		return errors.Wrapf(err, "xtrabackup err: %s\n xbcloud out: %s\n xbcloud err: %s",
+			errBackupBuffer.String(), outBuffer.String(), errCloudBuffer.String())
+	}
+
+	if xbcloudCmd != nil {
+		xbcloudCmd.Stdin = xtrabackupStdout
+		xbcloudCmd.Stdout = &outBuffer
+		xbcloudCmd.Stderr = &errCloudBuffer
+		if err := xbcloudCmd.Start(); err != nil {
+			return wrapError(err)
+		}
+	}
+
+	if err := xtrabackupCmd.Start(); err != nil {
+		return wrapError(err)
+	}
+
+	if xbcloudCmd != nil {
+		if err := xbcloudCmd.Wait(); err != nil {
+			return wrapError(err)
```

Comment:
so we don't wait until `xtrabackupCmd` finished, right? let's create cancellable context and cancel it once we get error

Reply:
Context cancel handled by jobsRunner. So once job return error, jobRunner will kill its context.